Handle terminfo envvar

ijc opened the pull request:
As reported in https://github.com/docker/docker/issues/32400 (née https://github.com/docker/for-mac/issues/1515) setting `$TERMINFO` leads to a panic because `OpenTermInfo` has no code to handle this case.

This PR cleans up a bunch of stuff in that function and adds the missing case for when `$TERMINFO` is set.

NB I also include https://github.com/ijc25/Gotty/commit/f56504781b6b0704731705d2d411076a24b04c0f which is also in #1 so that `go fmt -s` is clean.

Commit:
Run `gofmt -s -w *.go`, stripping a single trailing space.

Commit:
Use path/filepath for path construction.

Signed-off-by: Ian Campbell <ian.campbell@docker.com>

Commit:
Simplify searching default terminfo paths.

- Checking if the file is unreadable is not needed, just try to open and read
  and handle the error.
- Upon success just return the result rather than breaking from the loop and
  falling through, which requires more complex error handling.
- Upon overall failure return the error directly. There is no need to
  fmt.Sprintf the static string argument to errors.New()

This does not yet correctly handle the case where $TERMINFO contains a value.

Signed-off-by: Ian Campbell <ian.campbell@docker.com>

Commit:
Do not search $HOME/.terminfo if $HOME is not set

Signed-off-by: Ian Campbell <ian.campbell@docker.com>

Commit:
Handle $TERMINFO != ""

https://linux.die.net/man/5/terminfo says:
> If the environment variable TERMINFO is set, it is interpreted as the
> pathname of a directory containing the compiled description you are working
> on. Only that directory is searched.

I'd have preferred a more authoritative/OS neutral reference (such as
http://pubs.opengroup.org/onlinepubs/7908799/xcurses/terminfo.html) but was
unable to find one.

Signed-off-by: Ian Campbell <ian.campbell@docker.com>

thaJeztah commented on gotty.go:
```diff
@@ -21,33 +22,29 @@ import (
-	termloc := os.Getenv("TERMINFO")
-	if len(termloc) == 0 {
+	if termloc := os.Getenv("TERMINFO"); len(termloc) > 0 {
+		path := filepath.Join(termloc, string(termName[0]), termName)
```
This is only used on Linux, so perhaps `path.Join()` instead of `filepath.Join()`?

ijc replied:
I think using the thing which is semantically correct everywhere is the right thing to do even if it isn't strictly necessary for a given case. Unless there is some massive performance overhead of `filepath` vs `path` but I doubt that.

thaJeztah replied:
My worry is that when running in, e.g. `git bash`, I _think_ it would pick back-slashes as a separator, but the environment expects forward slashes. It's a change in behaviour at least (existing code hard-codes `/`, and everything else assumes forward slashes)

@jhowardmsft @rneugeba to the rescue on what is used in that situation :blush:

thaJeztah replied:
One more thing; `OpenTermInfo("")` will panic. Given that this function is exported, I suggest to add a check for this

```go
if termName == "" {
    nil, errors.New("No terminfo file(-location) found")
}
```

ijc replied:
Ah, because of the `termName[0]`. Will push another fixup, thanks

thaJeztah reviewed (changes requested):
one question, but LGTM otherwise

Commit:
Use path.Join rather than filepath.Join

Using filepath switches from `/` to `\` separators on some platforms which
might be unexpected, so defer such a change to another time/discussion.

Signed-off-by: Ian Campbell <ian.campbell@docker.com>

thaJeztah commented:
See you just pushed; I just added another comment https://github.com/Nvveen/Gotty/pull/2#discussion_r110135030 :joy:

Commit:
Avoid panic on termName[0] if termName == ""

Signed-off-by: Ian Campbell <ian.campbell@docker.com>

thaJeztah reviewed (approved):
LGTM

(but would squishy squashy some of those commits :wink:)

ijc commented:
I'll leave squashing until @Nvveen requests it.

andriisoldatenko commented:
@ijc @thaJeztah @Nvveen any luck to merge this?

thaJeztah commented:
For Docker/Moby, we replaced this dependency with https://github.com/morikuni/aec, which is slightly more maintained, and was already in use by some other tools (see https://github.com/moby/moby/pull/38398)
